content/blog/anarchos/2024-04-09_an_odissey_to_port_compcert.md

```diff
@@ -0,0 +1,94 @@
+- `Compcert` is installed from its `Coq` sources. So i need `Coq`.
+- `Coq` is installed from sources or as an OPAM package. OPAM being the *Ocaml PAckage Manager*.
+- `opam` needs *OCaml* anyway. So i need *OCaml*.
+- `OCaml` is installable through *OPAM* or from sources.
+
+### The easiest part
+
+Porting OCaml was not too hard: copying most of Linux configuration in the `configure{.ac}` files work as is. 
+Don't forget to add some platform detection in `config.guess`, and some libraries tweaking:
+- -lm is not necessary
+- -lbsd is mandatory to use `getentropy`
+- -lnetwork to use... well networks :)
+
+### The easy part
+
+Compile `opam`.
+This one runs smoothly as soon as you have a working Ocaml compiler : `./configure && make && make install` does the trick.
+Don't forget to add the ususal ` --prefix ~/config/non-packaged` to avoid installing in `/usr/...`.
+
+### The normal part
+
+Try to install `Coq` : 
+`opam install coq` --> The coq package relies on the (ocaml) package `lablgtk3`. This one is used in `coqide` an IDE for Coq.
+The critical component used there is `gtksourceview3`, used to render Coq source files with syntax coloring. 
+
+### The journey begins...
+For months or even more than a year, i could not get past this lablgtk3 package, cause *Haiku* only gets `gtksourceview-4` in `haikuports/x11-libs/gtksourceview`.
+So i tried to trick lablgtk3 to compile with gtksourceview-3 files symlinked to gtksourceview-4 files, and a bit of renaming 3 <-> 4 here and there. 
+It worked but `coqide` crashed pretty soon when used due to this poor setup.
+
+So i decided to take the other way : port `gtksourceview3` properly.
+I achieved that last month, the PR is in phase of polishing, but was enough for me to carry on the work further.
+
+### The journey continues
+During my tasks, i encounter some annoying behabviour by `opam` :
+the `opam install` command uses the _Ocaml_ *dune* build system. But in some cases *dune* fails with a rather cryptic (at this time for me, not for more experimented Haiku developers than me) message :
+`execve() : Operation not supported`.
+At first, i thought it was because of the spawn submodule of dune. So i tried for months to do a pull request (PR) on Haiku, to kill other threads and do a fork() from another thread of the team than the main thread. I could never finish that because the main thread is very special, so you can't easily do a `fork()` from a non-main thread. But indeed it was in a file which calls execve (which is incompatible with `fork()` on Haiku if the team has more than one thread. So i disabled the call to execve(), as for the Win32 platform.
+
+### Finally, land is in sight
+When i am able to have a working `dune`, i can compile Coq : <img src="/files/blog/anarchos/coqide.png"/>
+
+### Ulysses goes home
+with coq available, i was able to compile `ccomp`, the CompCert certified compiler.
+First test : `
+[Rocker] ~/OCAML/Compcert> ccomp
+ccomp: error: no input file
+1 error detected.
+`
```

Comment:
Seems like using single backticks puts the text onto one single line (at least on GitHub), like this:
`[Rocker] ~/OCAML/Compcert> ccomp
ccomp: error: no input file
1 error detected.`

````suggestion
First test:
```
[Rocker] ~/OCAML/Compcert> ccomp
ccomp: error: no input file
1 error detected.
```
````